src/oic/oic/provider.py

```diff
@@ -2049,3 +1882,471 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+
+        :param cinfo: Client information
+        :param sub: Subject identifier
+        :param sid: The Issuer ID
+        :return: Tuple with logout URI and signed logout token
+        """
+        try:
+            back_channel_logout_uri = cinfo["backchannel_logout_uri"]
+        except KeyError:
+            return None
+
+        # always include sub and sid so I don't check for
+        # backchannel_logout_session_required
+
+        payload = {
+            "sub": sub,
+            "sid": sid,
+            "events": {BACK_CHANNEL_LOGOUT_EVENT: {}},
+            "jti": uuid.uuid4().hex,
+        }
+
+        try:
+            alg = cinfo["id_token_signed_response_alg"]
+        except KeyError:
+            alg = self.capabilities["id_token_signing_alg_values_supported"][0]
+
+        _jws = JWT(self.keyjar, iss=self.name, lifetime=86400, sign_alg=alg)
+        sjwt = _jws.pack(aud=cinfo["client_id"], **payload)
+
+        return back_channel_logout_uri, sjwt
+
+    def clean_sessions(self, usids: List[str]):
+        """
+        Remove Session IDs from the session DB.
+
+        :param usids: List of session IDs
+        """
+        _sdb = self.sdb
+        # Clean out all sessions
+        for sid in usids:
+            del _sdb[sid]
+
+    def logout_info_for_all_clients(
+        self, uid: Optional[str] = "", sid: Optional[str] = ""
+    ) -> Dict:
+        """
+        Collect information necessary to logout one user from all clients he/she has been using.
+
+        One of uid and sid MUST be provided. If uid is provided sid is ignored.
+        NO changes are made to the session DB.
+        No logout is actually performed
+        :param uid: User ID
+        :param sid: Session ID
+        :return: Dictionary with logout information
+        """
+        if not uid:
+            if not sid:
+                raise ParameterError("One of uid and sid MUST be provided")
+            else:
+                uid = self.sdb.get_uid_by_sid(sid)
+
+        # Find all the session IDs this user has gotten
+        usids = session_get(self.sdb, "uid", uid)
+        # Find all RPs this user has logged it from
+        _client_sid = {}
+        for usid in usids:
+            _client_sid[self.sdb[usid]["client_id"]] = usid
+
+        # Front-/Backchannel logout ?
+        _cdb = self.cdb
+        _iss = self.name
+        bc_logouts = {}
+        fc_iframes = {}
+        for _cid, _csid in _client_sid.items():
+            if "backchannel_logout_uri" in _cdb[_cid]:
+                _sub = self.sdb[_csid]["sub"]
+                _spec = self.do_back_channel_logout(_cdb[_cid], _sub, _csid)
+                if _spec:
+                    bc_logouts[_cid] = _spec
+            elif "frontchannel_logout_uri" in _cdb[_cid]:
+                # Construct an IFrame
+                _iframe = self.do_front_channel_logout_iframe(_cdb[_cid], _iss, _csid)
+                if _iframe:
```

Comment:
This is missing tests for False case, where do_front_channel_logout returns empty _iframe. Also it should probably be an explicit test for None.

Reply:
Same as above, when the condition can never be False, it shouldn't be here.